### PR TITLE
load react app only if overlayPermission is resolved

### PR DESCRIPTION
**Pull request by ivanbogdanov**

### CODE REVIEW

File: libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/ui/SalesforceReactActivity.java [hunk 1]
```diff
@@ -276,6 +284,25 @@ protected boolean shouldReactBeRunning(){
...
-        SalesforceReactActivity.this.getReactNativeHost().getReactInstanceManager().createReactContextInBackground();
+        if(shouldReactBeRunning()){
+            SalesforceReactActivity.this.getReactNativeHost().getReactInstanceManager().createReactContextInBackground();
+        }
+    }
+
+    private boolean shouldAskOverlayPermission(){
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            if(SalesforceReactActivity.this.getReactNativeHost().getReactInstanceManager().getDevSupportManager().getDevSupportEnabled()){
+                if (!Settings.canDrawOverlays(this)) {
```
Comment (wmathurin) [inline libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/ui/SalesforceReactActivity.java]: Where is the code that prompt the user to grant that permission?

Comment (wmathurin) [inline libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/ui/SalesforceReactActivity.java]: So it happens when the delegate is created. So what's the user experience like, they see the popup and then when they go back, the react activity is resumed, shouldReactBeRunning is true, and react loads?
What if they answered no. Are you stuck with a grey screen?
When shouldAskOverlayPermission returns false, should we ask for permission?

**Comment by ivanbogdanov** [comment]
Uh good point good point



On Dec 7, 2016, at 2:16 PM, Wolfgang Mathurin <notifications@github.com>
wrote:

*@wmathurin* commented on this pull request.
------------------------------

In
libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/ui/SalesforceReactActivity.java
<https://github.com/forcedotcom/SalesforceMobileSDK-Android/pull/1319>:

> @@ -276,6 +284,25 @@ protected boolean shouldReactBeRunning(){

     protected void restartReactNativeApp(){
         SalesforceReactActivity.this.getReactNativeHost().getReactInstanceManager().destroy();
-        SalesforceReactActivity.this.getReactNativeHost().getReactInstanceManager().createReactContextInBackground();
+        if(shouldReactBeRunning()){
+            SalesforceReactActivity.this.getReactNativeHost().getReactInstanceManager().createReactContextInBackground();
+        }
+    }
+
+    private boolean shouldAskOverlayPermission(){
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            if(SalesforceReactActivity.this.getReactNativeHost().getReactInstanceManager().getDevSupportManager().getDevSupportEnabled()){
+                if (!Settings.canDrawOverlays(this)) {

So it happens when the delegate is created. So what's the user experience
like, they see the popup and then when they go back, the react activity is
resumed, shouldReactBeRunning is true, and react loads?
What if they answered no. Are you stuck with a grey screen?
When shouldAskOverlayPermission returns false, should we ask for permission?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/forcedotcom/SalesforceMobileSDK-Android/pull/1319>, or mute
the thread

**Comment by ivanbogdanov** [comment]
Handle cancelled permission overlay dialog situation

**Comment by ivanbogdanov** [comment]
@wmathurin @bhariharan

File: libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/ui/SalesforceReactActivity.java [hunk 2]
```diff
@@ -276,6 +289,54 @@ protected boolean shouldReactBeRunning(){
...
+    }
+
+    private void loadReactAppOnceIfReady() {
+        if(reactActivityDelegate != null ){
+            reactActivityDelegate.loadReactAppOnceIfReady(getMainComponentName());
+        }
+    }
+
+    private void showPermissionWarning(){
+        if(overlayPermissionRequiredDialog == null){
```
Comment (ivanbogdanov) [inline libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/ui/SalesforceReactActivity.java]: The actual overlay dialog is triggered here:
https://github.com/facebook/react-native/blob/0.35-stable/ReactAndroid/src/main/java/com/facebook/react/ReactActivityDelegate.java#L83

Yes. If overlay perm dialog is closed without confirming overlay draw perm: this information dialog is displayed: (I didn't want to duplicate the existing code for this corner case) 
When a user presses 'Continue' button on this informational dialogue: the activity gets recreated and  ReactActivityDelegate opens Overlay Perm screen again.

Comment (bhariharan) [inline libs/SalesforceReact/src/com/salesforce/androidsdk/reactnative/ui/SalesforceReactActivity.java]: 👍

**Comment by ivanbogdanov** [comment]
Looks like the build and tests are ok: there is some CI job issue there

**Comment by bhariharan** [comment]
Yup, go ahead and merge if tests are passing locally.

**Pull request merged**